cloud/ec2.go

```
@@ -709,46 +709,60 @@ func (m *ec2Manager) GetInstanceStatuses(ctx context.Context, hosts []host.Host)
+		InstanceIds: hostsToCheck,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "describing instances")
+	}
+	if err = validateEc2DescribeInstancesOutput(out); err != nil {
+		return nil, errors.Wrap(err, "invalid describe instances response")
+	}
+
+	reservationsMap := map[string]types.Instance{}
+	for i := range out.Reservations {
+		reservationsMap[*out.Reservations[i].Instances[0].InstanceId] = out.Reservations[i].Instances[0]
+	}
+
+	hostsToCache := make([]hostInstancePair, 0, len(hostsToCheck))
+	hostIDsToCache := make([]string, 0, len(hostsToCheck))
+	hostToStatusMap := make(map[string]CloudStatus, len(hostsToCheck))
+	for i := range hostsToCheck {
+		hostID := hostsToCheck[i]
+		instance, ok := reservationsMap[hostID]
+		if !ok {
+			hostToStatusMap[hostID] = StatusNonExistent
+			continue
 		}
-		for i := range hostsToCheck {
-			instance, ok := reservationsMap[hostsToCheck[i]]
-			if !ok {
-				hostToStatusMap[hostsToCheck[i]] = StatusNonExistent
-				continue
-			}
-			status := ec2StatusToEvergreenStatus(instance.State.Name)
-			if status == StatusRunning {
-				// cache instance information so we can make fewer calls to AWS's API
-				if err = cacheHostData(ctx, m.env, instanceIdToHostMap[hostsToCheck[i]], &instance, m.client); err != nil {
-					return nil, errors.Wrapf(err, "caching EC2 host data for host '%s'", hostsToCheck[i])
-				}
-			}
-			hostToStatusMap[instanceIdToHostMap[hostsToCheck[i]].Id] = status
+		status := ec2StatusToEvergreenStatus(instance.State.Name)
+		if status == StatusRunning {
+			hostsToCache = append(hostsToCache, hostInstancePair{
+				host:     instanceIdToHostMap[hostID],
+				instance: &instance,
+			})
+			hostIDsToCache = append(hostIDsToCache, hostID)
 		}
+		hostToStatusMap[hostID] = status
 	}
 
+	// Cache instance information so we can make fewer calls to AWS's API.
+	grip.Error(message.WrapError(cacheAllHostData(ctx, m.env, m.client, hostsToCache), message.Fields{
```

Why not return an error on failure as we were doing before?